### PR TITLE
Tunnel Termination ACL VPP Plugin

### PR DESCRIPTION
**Pull request by AkeelAli**
Overview:
--------
This PR introduces a custom VPP plugin named `tunterm_acl` into Sonic-VPP. This plugin provides the required dataplane ACL functionality to support Smart Switch HA. In particular, it allows v4-vxlan decap'ed packets to be redirected based on their inner DST IP.

Changes:
--------
1. Adds the `tunterm_acl` plugin source code (including tests and docs) under `platform/vpp/vppbld/plugins/`
2. Updates the `vppbld` Makefile to allow for custom plugins to be copied to & built with VPP
3. Enables the `tunterm_acl` plugin in the `startup.conf` files
4. Adds a change in the `vpp.patch` file to export `vxlan_main`

Plugin:
-----
Details of the `tunterm_acl` plugin implementation can be found in the [README.rst](https://github.com/yue-fred-gao/sonic-platform-vpp/blob/tunterm_acl_plugin/platform/vpp/vppbld/plugins/tunterm_acl/docs/README.rst) file.

### CODE REVIEW

File: platform/vpp/vppbld/Makefile [hunk 1]
```diff
@@ -23,6 +23,7 @@ UID = $(shell id -u)
 GUID = $(shell id -g)
 
 VPP_REPO_DIR = repo
+VPP_CUSTOM_PLUGINS_DIR = ../plugins
```
Comment (yue-fred-gao) [inline platform/vpp/vppbld/Makefile:26]: I think we should move plugins directory into vppbld, which is for building vpp. So everything for vpp build is under the same directory.

Comment (AkeelAli) [reply platform/vpp/vppbld/Makefile]: Sounds good. I moved the plugins under vppbld in https://github.com/sonic-net/sonic-platform-vpp/pull/114/commits/935bb7992331e939ac92348a3ed225e1873aaa34.

File: platform/vpp/plugins/tunterm_acl/tunterm_acl_api.c [hunk 2]
```diff
@@ -0,0 +1,340 @@
...
+  if (rv != 0) {
+    clib_warning ("tunterm_acl_redirect_clear failed");
+    return rv;
+  }
+
+  /* Now add the new stuff */
+  for (int i = 0; i < count; i++) {
+    /* (1) Process Route */
+    fib_route_path_t *paths_ = 0;
+    u8 n_paths = 1; //rules[i].n_paths;
```
[the remark this answers is not in the document]
Comment (AkeelAli) [reply platform/vpp/plugins/tunterm_acl/tunterm_acl_api.c]: Yes. I updated the comment and the README file to add this as enhancement/TODO item. It's also mentioned in the api file.

File: platform/vpp/plugins/tunterm_acl/tunterm_acl_api.c [hunk 3]
```diff
@@ -0,0 +1,340 @@
...
+        vec_free (paths_);
+        return rv;
+      }
+      vec_add1 (paths_, path);
+    }
+
+    /* (2) Process DST IP */
+    if (is_ipv6 != rules[i].dst.af) {
+      return VNET_API_ERROR_INVALID_VALUE_3;
+      vec_free (paths_);
```
[the remark this answers is not in the document]
Comment (AkeelAli) [reply platform/vpp/plugins/tunterm_acl/tunterm_acl_api.c]: Good catch. Fixed in latest commit.

File: platform/vpp/plugins/tunterm_acl/tunterm_acl_decap.c [hunk 4]
```diff
@@ -0,0 +1,627 @@
...
+				 sizeof (ip6_header_t) +
+				 sizeof (udp_header_t));
+
+	exit1:
+	  vlib_validate_buffer_enqueue_x2 (vm, node, next_index,
+					   to_next, n_left_to_next,
+					   bi0, bi1, next0, next1);
+	}
+
+      while (n_left_from > 0 && n_left_to_next > 0)
```
[the remark this answers is not in the document]
Comment (AkeelAli) [reply platform/vpp/plugins/tunterm_acl/tunterm_acl_decap.c]: These loops implement the single/dual/quad loop packet processing pattern used in VPP nodes for performance optimization. See this link for more details: https://fd.io/docs/vpp/v2005/gettingstarted/developers/vnet#effective-graph-dispatch-function-coding. You'll notice my tunterm-acl node doesn't have these loops yet as performance wasn't a concern/priority for us. It's part of the enhancements lists in the README file (the node can be evolved to align with this model if needed).

**Review by yue-fred-gao — APPROVED** [review]
Looks good from SONIC side. Let's wait for Abdel to review from vpp

**Review by abdbaig — CHANGES_REQUESTED** [review]
If not already done, can you use the plugin source code within vpp and run the following to make sure all the code is validated against vpp guidelines:

make checkstyle
make checkstyle-api
make chekcstyle-python

The code may have to committed in local branch for the scripts to work.

File: platform/vpp/vppbld/plugins/tunterm_acl/docs/README.rst [hunk 5]
```diff
@@ -0,0 +1,66 @@
...
+Tunterm ACL Plugin README
+=======================
+
+Overview
+--------
+The Tunterm ACL plugin is a self-contained plugin that applies ACL after Tunnel Decapsulation.
+
+It is currently designed to support a single specific use-case:
+
+IPv4 VxLAN tunnel termination and classification based on inner DST IPv4/6 fields, followed by a redirect action using ip4/6-rewrite.
```
[the remark this answers is not in the document]
Comment (AkeelAli) [reply platform/vpp/vppbld/plugins/tunterm_acl/docs/README.rst]: Clarified that this is a redirect via a VPP Fib path (as declared in the API using vl_api_fib_path_t)

File: platform/vpp/vppbld/plugins/tunterm_acl/tunterm_acl_api.c [hunk 6]
```diff
@@ -0,0 +1,340 @@
...
+
+    ip46_address_t dst;
+    ip_address_decode (&rules[i].dst, &dst);
+
+    u8 *match_vec = 0;
+    clib_memset (mask, 0, sizeof (mask));
+
+    if (is_ipv6) {
+      for (int i = 0; i < 16; i++) {
+        mask[8 + i + skip * CLASSIFY_TABLE_VECTOR_SIZE] = dst.ip6.as_u8[i];
```
[the remark this answers is not in the document]
Comment (AkeelAli) [reply platform/vpp/vppbld/plugins/tunterm_acl/tunterm_acl_api.c]: Removed in latest commit (no longer offset, masking directly from start of DIP)

File: platform/vpp/vppbld/plugins/tunterm_acl/tunterm_acl_api.c [hunk 7]
```diff
@@ -0,0 +1,340 @@
...
+      }
+    } else {
+      for (int i = 0; i < 4; i++) {
+        mask[i + skip * CLASSIFY_TABLE_VECTOR_SIZE] = dst.ip4.data[i];
```
[the remark this answers is not in the document]
Comment (AkeelAli) [reply platform/vpp/vppbld/plugins/tunterm_acl/tunterm_acl_api.c]: Removed in latest commit (no longer skipping vectors, masking directly from start of DIP)

File: platform/vpp/vppbld/plugins/tunterm_acl/tunterm_acl_node.c [hunk 8]
```diff
@@ -0,0 +1,204 @@
...
+
+            b0 = vlib_get_buffer(vm, bi0);
+            sw_if_index0 = vnet_buffer(b0)->sw_if_index[VLIB_RX];
+
+            u8* etype = &(b0->data[b0->current_data + sizeof(vxlan_header_t) + 2*sizeof(mac_address_t)]);
+            u16 ethertype = (etype[0] << 8) | etype[1];
+
+            if (ethertype == 0x86DD) {
+                table_index = tunterm_acl_main.classify_table_index_by_sw_if_index_v6[sw_if_index0];
+            } else if (ethertype == 0x0800) {
```
[the remark this answers is not in the document]
Comment (AkeelAli) [reply platform/vpp/vppbld/plugins/tunterm_acl/tunterm_acl_node.c]: Added support for vlan tagged inner packet along with testing.

**Review by abdbaig — APPROVED** [review]

**Pull request merged**